cache/cache.go

```diff
@@ -0,0 +1,43 @@
+package cache // import "bosun.org/cache"
+
+import (
+	"bosun.org/_third_party/github.com/golang/groupcache/lru"
+	"bosun.org/_third_party/github.com/golang/groupcache/singleflight"
+	"sync"
+)
+
+type Cache struct {
+	g singleflight.Group
+
+	sync.Mutex
+	lru *lru.Cache
+}
+
+type Result struct {
+	Val interface{}
+	Err error
+}
+
+func New() *Cache {
+	return &Cache{
+		lru: lru.New(100),
+	}
+}
+
+func (c *Cache) Get(key string, getFn func() (interface{}, error)) (interface{}, error) {
+	c.Lock()
+	result, ok := c.lru.Get(key)
+	c.Unlock()
+	if ok {
+		res := result.(Result)
+		return res.Val, res.Err
+	}
+	// our lock only serves to protect the lru.
+	// we can (and should!) do singleflight requests concurently
+	val, err := c.g.Do(key, getFn)
+	r := Result{val, err}
+	c.Lock()
+	c.lru.Add(key, r)
```

Comment:
This has the potential to dogpile adding entries to the cache, including locking c.lru. What about moving c.lru.Add into getFn? For example:

```
val, err := c.g.Do(key, func() (interface{}, error) {
  v, err := getFn()
  c.lru.Add(key, &Result{v, err})
  return v, err
})
```

And remove the c.Lock/Unlock calls.